File: accumulated_incremental.py
###################################################################
#                                                                 #
#                        Produced by:                             #
#                 Marcus Nobrega Junior gomes                     #
#              (marcusnobrega.engcivil@gmail.com)                 #
#                               &                                 #
#                 Luis Miguel Castillo Rápalo                     #
#                   (luis.castillo@unah.hn)                       #
#                       September 2021                            #
#                                                                 #
#       Last update : 11 september, 2021                          #
###################################################################

# Calculates accumulated and incremental variable, assuming the model's
# time-step and the timeseries time-step
# 09/30/2020

import numpy as np

# delta_p is the variable discretized into model's time-step

def accumulated_incremental(steps, variable_discretized, time_step_variable):
    delta_variable = np.zeros((int(steps)))
    accum_variable = np.cumsum(variable_discretized[0,:]*time_step_variable/60)

    for t in np.arange(steps):
        if t == 0:
            delta_variable[int(t)] = accum_variable[int(t)]
        elif t >= len(variable_discretized[0,:]):
            delta_variable[int(t)] = 0
        else:
            delta_variable[int(t)] = accum_variable[int(t)]-accum_variable[int(t)-1] #mm

    variable_intensity = delta_variable/(time_step_variable/60) # mm/h trhoughall time-steps
    delta_variable = np.round(delta_variable, 6) #ROUNDED VALUE, CAREFULL HERE!

    return delta_variable, variable_intensity

File: HydroPol2D_Functions/accumulated_incremental.m
function [accum_variable, delta_variable, variable_intensity] = accumulated_incremental(steps, variable_discretized, time_step_variable)
% Optimized version of accumulated_incremental
% Author: Marcus Nobrega Gomes Junior
% Last Updated: Optimized by ChatGPT - July 2025

% Inputs:
%   steps                - Number of time steps
%   variable_discretized - Matrix of variable values (rows = series, cols = time)
%   time_step_variable   - Time step in minutes

% Outputs:
%   accum_variable       - Accumulated variable over time
%   delta_variable       - Incremental change at each time step
%   variable_intensity   - Intensity (e.g., mm/h)

% Precompute constants
dt_hours = time_step_variable / 60;

% Preallocate
[n_series, n_times] = size(variable_discretized);
delta_variable = zeros(n_series, steps);

% Only compute up to min(steps, n_times)
valid_steps = min(steps, n_times);

% Cumulative sum for accumulation
accum_variable = cumsum(variable_discretized(:,1:valid_steps) * dt_hours, 2);

% Compute delta values efficiently
delta_variable(:,1:valid_steps) = [accum_variable(:,1), diff(accum_variable, 1, 2)];

% Fill remaining steps (if any) with zeros
if steps > n_times
    delta_variable(:, n_times+1:steps) = 0;
end

% Intensity in mm/h
variable_intensity = delta_variable / dt_hours;

end
